test/GPv2Signing/RecoverOrderSigner.t.sol

```diff
@@ -0,0 +1,208 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+pragma solidity ^0.8;
+
+import {Vm} from "forge-std/Test.sol";
+
+import {EIP1271Verifier, GPv2EIP1271, GPv2Order, GPv2Signing, IERC20} from "src/contracts/mixins/GPv2Signing.sol";
+
+import {Helper} from "./Helper.sol";
+import {Order} from "test/libraries/Order.sol";
+import {Sign} from "test/libraries/Sign.sol";
+
+contract RecoverOrderSigner is Helper {
+    using GPv2Order for GPv2Order.Data;
+
+    Vm.Wallet private trader;
+
+    constructor() {
+        trader = vm.createWallet("GPv2Signing.RecoverOrderFromTrade: trader");
+    }
+
+    function defaultOrder() internal returns (GPv2Order.Data memory) {
+        return GPv2Order.Data({
+            sellToken: IERC20(makeAddr("GPv2Sign.RecoverOrderSigner: default order sell token")),
+            buyToken: IERC20(makeAddr("GPv2Sign.RecoverOrderSigner: default order buy token")),
+            receiver: makeAddr("GPv2Sign.RecoverOrderSigner: default order receiver"),
+            sellAmount: 42 ether,
+            buyAmount: 13.37 ether,
+            validTo: type(uint32).max,
+            appData: keccak256("GPv2Sign.RecoverOrderSigner: app data"),
+            feeAmount: 1 ether,
+            kind: GPv2Order.KIND_SELL,
+            partiallyFillable: false,
+            sellTokenBalance: GPv2Order.BALANCE_ERC20,
+            buyTokenBalance: GPv2Order.BALANCE_ERC20
+        });
+    }
+
+    function test_should_recover_signing_address_for_all_supported_ECDSA_based_schemes() public {
+        GPv2Order.Data memory order = defaultOrder();
+
+        Sign.Signature memory eip712Signature = Sign.sign(vm, trader, order, GPv2Signing.Scheme.Eip712, domainSeparator);
+        address eip712Owner = executor.recoverOrderSignerTest(order, eip712Signature);
+        assertEq(eip712Owner, trader.addr);
+
+        Sign.Signature memory ethSignSignature =
+            Sign.sign(vm, trader, order, GPv2Signing.Scheme.EthSign, domainSeparator);
+        address ethSignOwner = executor.recoverOrderSignerTest(order, ethSignSignature);
+        assertEq(ethSignOwner, trader.addr);
+    }
+
+    function test_reverts_for_invalid_signing_schemes() public {
+        vm.expectRevert();
+        executor.recoverOrderSignerTest(
+            defaultOrder(), Sign.Signature({scheme: GPv2Signing.Scheme(uint8(42)), data: hex""})
```

Comment:
This test isn't achieving it's actual desired outcome as attempting to convert `uint8(42)` into `GPv2Signing.Scheme` causes the a revert with a panic at this specific point in the test before there's an actual revert on `recoverOrderSignerTest`, so it seems that this would require a low-level call to by-pass the solidity type checking.

Reply:
Huh, you're absolutely right, great find!
My expectation was that `expectRevert` catches a revert in the next call, but it just catches any revert that follows! So this is a successful test:
```solidity
vm.expectRevert("test revert");
revert("test revert");
```

Also, there are valid tests that cannot be compiled, for example:
```solidity
vm.expectRevert("test revert");
revert("test revert");
return;
```

Which fails compilation with:
```
Warning (5740): Unreachable code.
  --> test/GPv2Signing/RecoverOrderSigner.t.sol:54:9:
   |
54 |         return;
```

Reply:
I thought a bit about it and it's hard to meaningfully simulate this. We are testing a library, so we can't call something and see a revert because all of this is handled at a compiler level. So I ended up simplifying the test to just revert if a function tries to internally generate an invalid signing scheme. WDYT?

Reply:
Do we think that this value adds? I'd argue that because it can be tested in solidity, and the `uint8` conversion of the enum is bound by types within the compiler, it's safe for us to remove this test entirely.

Reply:
Fair enough. Test has been removed and PR description has been updated.